woocommerce/woocommerce-ios
# Order Creation: Sync product list from remote in Add Product screen

Part of: #5408

## Description

In Order Creation, the Add Product screen now syncs the product list from remote, with the following behavior:

- When the Add Product screen first opens, and no products exist in storage, a ghost list view appears. Once the first page of products are synced, the ghost list is replaced by a real list of products.
- When the Add Product screen first opens, and there are products in storage, the stored products appear immediately (with no ghost view).
- If there are more products to sync from remote, an infinite scroll indicator appears at the bottom of the list while the next page of products are syncing.

Several notes/caveats from #5589 still apply:

- For now we're only loading non-variable products. (Variable products will be implemented later.)
- The product list only shows published and private products. This matches core/web behavior, where draft and pending products can't be added to an order.
- For now we're only showing a placeholder product image. (Product images will be added in another PR.)

## Changes

* `AddProductToOrder`:
   * The `AddProductToOrder` view now switches on `SyncStatus` to display the appropriate view: a list of products if there are any products to show, an empty state if there are no products, a ghost list during the initial sync, and an empty view as a default (initial state).
   * When `AddProductToOrder` appears, the initial sync is triggered. Subsequent syncs are triggered when the last row of the product list appears.
* `AddProductToOrderViewModel`:
   * The list of products is now filtered in `products` instead of the results controller. This allows us to use the full list of products in the results controller to trigger the next page for syncing.
   * Adds a `SyncCoordinator` and necessary methods to sync products from remote, update the `syncStatus` before/during/after sync, and update the results controller after sync.
* `ProductRowViewModel` is now `Equatable` to be able to check when a product row is the last row in the list.
* Adds unit tests for the scrolling indicator and sync status changes.

## Testing

1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab and tap the + button. (If Simple Payments is also enabled, tap "Create order" in the bottom sheet that appears.)
4. On the New Order screen, tap the "Add product" button.
5. On the Add Product screen, confirm you see the expected behavior based on whether you have any products in storage and whether your store has any non-variable, published or private products to sync from remote.

Scenarios to consider:

* If you have not synced any products from remote yet (e.g. log in and immediately create a new order, without opening the Products tab) you should see the ghost list while the first sync is happening.
* If you have synced _all_ of your store's products from remote, the full list of non-variable, published or private products should be loaded in the product list.
* If you have synced _some_ but not all of your store's products from remote, when you scroll down you should see the scroll indicator (spinner) at the bottom of the list while more products are being synced.
* If you don't have any products on your store, you should see an empty view.

## Screenshots


https://user-images.githubusercontent.com/8658164/145828789-2d711050-c08b-445d-9741-62dbdd5bc2eb.mp4



Empty view|Loading view (first sync)|Product list
-|-|-
![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 37 19](https://user-images.githubusercontent.com/8658164/145828699-934864d6-582e-4df8-9973-aaf62d7fc77e.png)|![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 37 58](https://user-images.githubusercontent.com/8658164/145828709-8c5f172d-d1a3-4bc5-8786-84b029800f8e.png)|![Simulator Screen Shot - iPhone 13 Pro - 2021-12-13 at 12 38 01](https://user-images.githubusercontent.com/8658164/145828718-c13648ad-2b2b-4e5f-bb6e-ad74e42d78de.png)


## Submitter Checklist

Update release notes:
